counsel.el

```diff
@@ -6502,16 +6548,46 @@ specified by the `blddir' property."
              (compile cmd)
           (remove-hook 'compilation-start-hook #'counsel-compile--update-history))))))
 
+(defun counsel-compile-edit-command ()
+  "Insert current compile command into mini-buffer for editing.
+
+This mirrors the behaviour of `ivy-insert-current' but with specific
+handling for the counsel-compile metadata."
+  (interactive)
+  (delete-minibuffer-contents)
+  (let* ((end (and ivy--directory
+                   (ivy--dirname-p (ivy-state-current ivy-last))
+                   -1))
+         (cmd (ivy-state-current ivy-last))
+         (blddir (get-text-property 0 'blddir cmd)))
+    (when blddir
+      (setq counsel-compile--current-build-dir blddir))
+    (if (get-char-property 0 'cmd cmd)
+        (insert (substring-no-properties
+                 cmd 0 (next-single-property-change 0 'cmd cmd)))
+      (insert (substring-no-properties cmd)))))
+
+;; Currently the only thing we do is override ivy's default insert
+;; operation which doesn't include the metadata we want.
+(defvar counsel-compile-map
+  (let ((map (make-sparse-keymap)))
+    (define-key map (kbd "M-i") #'counsel-compile-edit-command)
```

Comment:
Would `[remap ivy-insert-current]` work? And would it be any better? (I don't know.)